Sources/PipeableSDK/PageLoadState.swift

```diff
@@ -39,11 +39,57 @@ class PageLoadState {
         signalLoadStateChange()
     }
 
-    /// Subscribe to the load state change.
-    /// - Parameter callback: The callback to be called when the load state
-    /// changes. The callback should return true if it wants to be removed from
-    /// the list of subscribers.
-    /// - Returns: A function that can be called to unsubscribe from the load
+    func waitForLoadStateChange(predicate: @escaping (_ state: LoadState, _ url: String?) -> Bool, timeout: Int) async throws {
+        // Since there is a potential race condition that can lead to double
+        // "resume" calls on the continuation, we need to ensure that the
+        // continuation is only resumed once. We guard this by running resumes
+        // in a queue and using a helper.
+
+        let resumeOnce = ResumeOnce()
+
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+            var timeoutTask: Task<Void, Never>?
+
+            let removeListener = self.subscribeToLoadStateChange { state, url, error in
+                if let error = error {
+                    resumeOnce.resume {
+                        continuation.resume(throwing: error)
+                    }
+
+                    // Clean up timer.
+                    timeoutTask?.cancel()
+
+                    return true
+                } else if predicate(state, url) {
+                    resumeOnce.resume {
+                        continuation.resume(returning: ())
+                    }
+
+                    // Clean up timer.
+                    timeoutTask?.cancel()
+                    return true
+                }
+
+                return false
+            }
+
+            timeoutTask = Task {
+                do {
+                    try await Task.sleep(nanoseconds: UInt64(timeout) * 1000000)
+                } catch {
+                    // If the sleep is cancelled, then we move to
```

Comment:
Yes. Will change it

Reply:
we move on, yes.

It's equivalent to just returning, too, since if the sleep is cancelled, then we should be in the case where the timeout task is being canceled, meaning we've handled the case already.